### PR TITLE
[Substrait to Velox] Support the conversion between SingularOrList and Velox IN expression

### PR DESCRIPTION
**Pull request by rui-mo**
SingularOrList ([link](https://github.com/facebookincubator/velox/blob/main/velox/substrait/proto/substrait/algebra.proto#L927-L930)) in Substrait is a specialized structure that is often used is a large list of possible values. In SQL, these are typically large IN lists. For single value expressions, these are a compact equivalent of `expression = value1 OR expression = value2 OR .. OR expression = valueN`. (Reference: [or-list-equality-expression](https://substrait.io/expressions/specialized_record_expressions/#or-list-equality-expression))
This PR supports the conversion between SingularOrList and Velox IN expression. The conversion process is similar as what is done in [InTest.cpp](https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/tests/InTest.cpp#L86-L90). I drew below flow chart, and hope it can help illustrate.

Screenshot from 2023-06-06 07-11-30

In this picture, `value` and `options` are two items of SingularOrList. The value is the variable, and the options are the possible values to be compared with. They represent two arguments for the creation of a Velox IN expression.

### CODE REVIEW

File: velox/substrait/VeloxToSubstraitExpr.h [hunk 1]
```diff
@@ -96,6 +96,13 @@ class VeloxToSubstraitExprConvertor {
       google::protobuf::Arena& arena,
       const std::shared_ptr<ConstantVector<ComplexType>>& constantVector);
 
+  /// Convert Velox IN expression to Substrait SingularOrList.
+  const ::substrait::Expression& toSubstraitSingularOrList(
+      google::protobuf::Arena& arena,
+      const std::vector<std::shared_ptr<const core::ITypedExpr>>& inputs,
```
Comment (zhejiangxiaomai) [inline velox/substrait/VeloxToSubstraitExpr.h:102]: maybe use TypedExprPtr to replace std::shared_ptr<const core::ITypedExpr>.

File: velox/substrait/SubstraitToVeloxExpr.cpp [hunk 2]
```diff
@@ -339,6 +340,46 @@ SubstraitVeloxExprConverter::toVeloxExpr(
...
+      singularOrList.options_size() > 0, "At least one option is expected.");
+  auto options = singularOrList.options();
+  std::vector<::substrait::Expression::Literal> literals;
+  literals.reserve(options.size());
+  for (const auto& option : options) {
+    VELOX_CHECK(option.has_literal(), "Literal is expected as option.");
+    literals.emplace_back(option.literal());
+  }
+
+  std::vector<std::shared_ptr<const core::ITypedExpr>> params;
```
Comment (zhejiangxiaomai) [inline velox/substrait/SubstraitToVeloxExpr.cpp:374]: maybe use  std::vector<TypedExprPtr> params;

File: velox/substrait/VeloxToSubstraitExpr.cpp [hunk 3]
```diff
@@ -452,6 +452,39 @@ VeloxToSubstraitExprConvertor::toSubstraitExpr(
   return *substraitFieldExpr;
 }
 
+const ::substrait::Expression&
+VeloxToSubstraitExprConvertor::toSubstraitSingularOrList(
+    google::protobuf::Arena& arena,
+    const std::vector<std::shared_ptr<const core::ITypedExpr>>& inputs,
```
Comment (zhejiangxiaomai) [inline velox/substrait/VeloxToSubstraitExpr.cpp:458]: const std::vector<TypedExprPtr>& inputs

File: velox/substrait/VeloxToSubstraitExpr.h (hunk 1 above)
Comment (rui-mo) [reply velox/substrait/VeloxToSubstraitExpr.h]: Fixed cases like this. Thanks for your review.

**Comment by rui-mo** [comment]
Hi @mbasmanova, would you like to review this PR? Thanks.

**Comment by rui-mo** [comment]
hi @mbasmanova @vibhatha, could you help review this PR when free? Thanks.

**Comment by rui-mo** [comment]
Hi @mbasmanova, would you like to review this PR? Or anyone else can help on that? Thanks.

**Comment by mbasmanova** [comment]
@rui-mo Rui, can you verify that the semantics of these two are exactly the same and explain how you did that in the PR description? Also, CI is failing. Please, take a look.

**Comment by rui-mo** [comment]
> @rui-mo Rui, can you verify that the semantics of these two are exactly the same and explain how you did that in the PR description? Also, CI is failing. Please, take a look.

Hi @mbasmanova, I updated the PR description to explain the semantics and the conversion process. CI was also fixed. Could you check again? Thank you.

**Comment by mbasmanova** [comment]
@rui-mo Rui, this description is very helpful. Thank you for putting it together. One question: why does the diagram has 2 paths? How do you decide which path to take?

File: velox/substrait/VeloxToSubstraitExpr.cpp [hunk 4]
```diff
@@ -462,6 +495,10 @@ const ::substrait::Expression& VeloxToSubstraitExprConvertor::toSubstraitExpr(
   auto inputs = callTypeExpr->inputs();
   auto& functionName = callTypeExpr->name();
 
+  if (functionName == "in") {
```
Comment (mbasmanova) [inline velox/substrait/VeloxToSubstraitExpr.cpp:498]: Hardcoding the name of the function is problematic. This function may be registered under a different name or a different function can be registered under the "in" name.

I wonder if a more robust approach is to introduce Options struct to specify various configurations for the converter and have one of these options be the name of the "IN" function.

File: velox/substrait/VeloxToSubstraitExpr.cpp [hunk 5]
```diff
@@ -452,6 +452,39 @@ VeloxToSubstraitExprConvertor::toSubstraitExpr(
...
 
+const ::substrait::Expression&
+VeloxToSubstraitExprConvertor::toSubstraitSingularOrList(
+    google::protobuf::Arena& arena,
+    const std::vector<core::TypedExprPtr>& inputs,
+    const RowTypePtr& inputType,
+    ::substrait::Expression* substraitExpr) {
+  ::substrait::Expression_SingularOrList* singularOrList =
+      substraitExpr->mutable_singular_or_list();
+  VELOX_CHECK(inputs.size() == 2, "Two arguments expected for IN.");
```
Comment (mbasmanova) [inline velox/substrait/VeloxToSubstraitExpr.cpp:463]: Use VELOX_CHECK_EQ for better error messages.

File: velox/substrait/VeloxToSubstraitExpr.cpp [hunk 6]
```diff
@@ -452,6 +452,39 @@ VeloxToSubstraitExprConvertor::toSubstraitExpr(
...
+  singularOrList->mutable_value()->MergeFrom(
+      toSubstraitExpr(arena, inputs.at(0), inputType));
+  auto constantExpr =
+      std::dynamic_pointer_cast<const core::ConstantTypedExpr>(inputs.at(1));
+  VELOX_CHECK_NOT_NULL(constantExpr);
+  VELOX_CHECK(
+      constantExpr->hasValueVector(),
+      "Value vector expected for constant expression.");
```
Comment (mbasmanova) [inline velox/substrait/VeloxToSubstraitExpr.cpp:471]: Should we also check the type of the expression? Should it be ARRAY(BIGINT())?

File: velox/substrait/SubstraitToVeloxExpr.h [hunk 7]
```diff
@@ -40,12 +40,12 @@ class SubstraitVeloxExprConverter {
       const RowTypePtr& inputType);
 
   /// Convert Substrait ScalarFunction into Velox Expression.
-  std::shared_ptr<const core::ITypedExpr> toVeloxExpr(
+  core::TypedExprPtr toVeloxExpr(
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxExpr.h:43]: Thank you for refactor. It would be nice to submit it as a separate PR to keep changes in this PR small.

File: velox/substrait/SubstraitToVeloxExpr.cpp [hunk 8]
```diff
@@ -339,11 +338,50 @@ SubstraitVeloxExprConverter::toVeloxExpr(
...
+  params.emplace_back(literalsToConstantExpr(literals));
+  return std::make_shared<const core::CallTypedExpr>(
+      BOOLEAN(), std::move(params), "in");
+}
+
+core::TypedExprPtr SubstraitVeloxExprConverter::toVeloxExpr(
     const ::substrait::Expression& substraitExpr,
     const RowTypePtr& inputType) {
-  std::shared_ptr<const core::ITypedExpr> veloxExpr;
+  core::TypedExprPtr veloxExpr;
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxExpr.cpp:384]: Is this variable used? If not, let's remove it.

File: velox/substrait/SubstraitToVeloxExpr.cpp [hunk 9]
```diff
@@ -339,11 +338,50 @@ SubstraitVeloxExprConverter::toVeloxExpr(
...
+  // Wrap the array vector into constant vector.
+  auto constantVector =
+      BaseVector::wrapInConstant(1 /*length*/, 0 /*index*/, arrayVector);
+  return std::make_shared<const core::ConstantTypedExpr>(constantVector);
+}
+
+core::TypedExprPtr SubstraitVeloxExprConverter::toVeloxExpr(
+    const ::substrait::Expression::SingularOrList& singularOrList,
+    const RowTypePtr& inputType) {
+  VELOX_CHECK(
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxExpr.cpp:362]: VELOX_CHECK_GT

File: velox/substrait/SubstraitToVeloxExpr.cpp [hunk 10]
```diff
@@ -339,11 +338,50 @@ SubstraitVeloxExprConverter::toVeloxExpr(
...
+    literals.emplace_back(option.literal());
+  }
+
+  std::vector<core::TypedExprPtr> params;
+  params.reserve(2);
+  // First param is the value, second param is the list.
+  params.emplace_back(toVeloxExpr(singularOrList.value(), inputType));
+  params.emplace_back(literalsToConstantExpr(literals));
+  return std::make_shared<const core::CallTypedExpr>(
+      BOOLEAN(), std::move(params), "in");
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxExpr.cpp:378]: Same comment about hard-coding "in".

**Comment by rui-mo** [comment]
> @rui-mo Rui, this description is very helpful. Thank you for putting it together. One question: why does the diagram has 2 paths? How do you decide which path to take?

@mbasmanova Sorry for the misleading. They are not two paths, but two items of  below `SingularOrList`.  The value is the variable, and the options are the possible values to be compared with. They are two arguments for the creation of a Velox IN expression.

```
  message SingularOrList {
    Expression value = 1;
    repeated Expression options = 2;
  }
```

File: velox/substrait/SubstraitToVeloxExpr.cpp [hunk 11]
```diff
@@ -339,11 +338,50 @@ SubstraitVeloxExprConverter::toVeloxExpr(
...
+  return std::make_shared<const core::ConstantTypedExpr>(constantVector);
+}
+
+core::TypedExprPtr SubstraitVeloxExprConverter::toVeloxExpr(
+    const ::substrait::Expression::SingularOrList& singularOrList,
+    const RowTypePtr& inputType) {
+  VELOX_CHECK(
+      singularOrList.options_size() > 0, "At least one option is expected.");
+  auto options = singularOrList.options();
+  std::vector<::substrait::Expression::Literal> literals;
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxExpr.cpp:365]: Can we avoid creating this vector and create a vector of TypedExprPtr instead?

File: velox/substrait/tests/VeloxSubstraitRoundTripTest.cpp [hunk 12]
```diff
@@ -402,6 +402,16 @@ TEST_F(VeloxSubstraitRoundTripTest, topNTwoKeys) {
       "SELECT * FROM tmp WHERE c0 > 15 ORDER BY c0 NULLS FIRST, c1 DESC NULLS LAST LIMIT 10");
 }
 
+TEST_F(VeloxSubstraitRoundTripTest, in) {
+  auto vectors = makeVectors(10, 4, 2);
+  createDuckDbTable(vectors);
+  auto plan = PlanBuilder(pool_.get())
+                  .values(vectors)
+                  .filter("c0 in (1, 2, 3)")
```
Comment (mbasmanova) [inline velox/substrait/tests/VeloxSubstraitRoundTripTest.cpp:410]: Let's add tests for all supported types of IN values.

**Comment by mbasmanova** [comment]
> They are not two paths, but two items of below `SingularOrList`. The value is the variable, and the options are the possible values to be compared with. They are two arguments for the creation of a Velox IN expression.

@rui-mo Got it. Thank you for clarifying. Perhaps, update the PR description for the benefit of future readers.

File: velox/substrait/VeloxToSubstraitExpr.cpp (hunk 5 above)
Comment (rui-mo) [reply velox/substrait/VeloxToSubstraitExpr.cpp]: Fixed.

File: velox/substrait/VeloxToSubstraitExpr.cpp (hunk 6 above)
Comment (rui-mo) [reply velox/substrait/VeloxToSubstraitExpr.cpp]: Added type check for ARRAY, but the element type can be of other type except for BIGINT.

File: velox/substrait/SubstraitToVeloxExpr.h (hunk 7 above)
Comment (rui-mo) [reply velox/substrait/SubstraitToVeloxExpr.h:43]: Sure. Submitted https://github.com/facebookincubator/velox/pull/5177.

File: velox/substrait/SubstraitToVeloxExpr.cpp (hunk 8 above)
Comment (rui-mo) [reply velox/substrait/SubstraitToVeloxExpr.cpp]: Fixed.

File: velox/substrait/SubstraitToVeloxExpr.cpp (hunk 9 above)
Comment (rui-mo) [reply velox/substrait/SubstraitToVeloxExpr.cpp]: Fixed.

File: velox/substrait/SubstraitToVeloxExpr.cpp (hunk 11 above)
Comment (rui-mo) [reply velox/substrait/SubstraitToVeloxExpr.cpp]: Fixed.

File: velox/substrait/SubstraitToVeloxExpr.cpp [hunk 13]
```diff
@@ -339,11 +338,50 @@ SubstraitVeloxExprConverter::toVeloxExpr(
...
+    const std::vector<::substrait::Expression::Literal>& literals) {
+  std::vector<variant> variants;
+  variants.reserve(literals.size());
+  VELOX_CHECK_GE(literals.size(), 0, "List should contain at least one item.");
+  for (const auto& literal : literals) {
+    variants.emplace_back(toVeloxExpr(literal)->value());
+  }
+  auto varArray = variant::array(variants);
+  ArrayVectorPtr arrayVector =
+      core::variantArrayToVector(varArray.inferType(), varArray.array(), pool_);
```
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/substrait/SubstraitToVeloxExpr.cpp]: Removed the use of inferType(), and use element type instead.

File: velox/substrait/VeloxToSubstraitExpr.cpp (hunk 4 above)
Comment (rui-mo) [reply velox/substrait/VeloxToSubstraitExpr.cpp:498]: Understand the issue here. I'm having two questions here:

- How do we know the exact `callTypeExpr->name()` of IN expression in Velox? It could vary according to the prefix.
- Can we record the prefix during registration and append it here?

**Comment by rui-mo** [comment]
Hi @mbasmanova, thanks for your review. I fixed most comments, and have some questions on the name and unit test. Could you take a look again? Thank you.

File: velox/substrait/VeloxToSubstraitExpr.cpp (hunk 4 above)
Comment (mbasmanova) [reply velox/substrait/VeloxToSubstraitExpr.cpp:498]: > How do we know the exact callTypeExpr->name() of IN expression in Velox?

We don't. Hence, I suggest we require that the user of the converter specify it.

File: velox/substrait/tests/VeloxSubstraitRoundTripTest.cpp (hunk 12 above)
Comment (mbasmanova) [reply velox/substrait/tests/VeloxSubstraitRoundTripTest.cpp]: > Current code goes to prestosql, and more types are supported in the IN function under sparksql. Do we need to use that?

It would be nice to test with both.

Comment (mbasmanova) [reply velox/substrait/tests/VeloxSubstraitRoundTripTest.cpp]: > For bigint and string, an error in duckdb parser is found at [DuckParser.cpp](https://github.com/facebookincubator/velox/blob/main/velox/duckdb/conversion/DuckParser.cpp#L373). The reason is the parser adds some casting to generate a literal of certain type.

Maybe this can be fixed (in a separate PR).

File: velox/substrait/VeloxToSubstraitExpr.cpp (hunk 4 above)
Comment (rui-mo) [reply velox/substrait/VeloxToSubstraitExpr.cpp]: I added prefix in the converter, because this is not just an issue for `IN`. The prefix of other scalar functions also needs to be considered.

File: velox/substrait/SubstraitToVeloxExpr.cpp (hunk 10 above)
Comment (rui-mo) [reply velox/substrait/SubstraitToVeloxExpr.cpp]: Fixed.

File: velox/substrait/tests/VeloxSubstraitRoundTripTest.cpp (hunk 12 above)
Comment (rui-mo) [reply velox/substrait/tests/VeloxSubstraitRoundTripTest.cpp]: > It would be nice to test with both.

Tests were added for IN of presto and spark sql function.

> Maybe this can be fixed (in a separate PR).

Added TODO in this PR. More debugging is needed for the conversion from string to expression lists in DuckDB parser.

**Comment by rui-mo** [comment]
Hi @mbasmanova, I fixed the hard-coded function name, and enabled spark `IN` tests. Could you review again? Thanks.

File: velox/substrait/VeloxToSubstraitExpr.cpp (hunk 4 above)
Comment (mbasmanova) [reply velox/substrait/VeloxToSubstraitExpr.cpp]: @rui-mo Rui, I'll take a look, but I don't think it is sufficient to specify he prefix. I imagine we need to specify full function names because it is not guaranteed that "in" name refers to the "right" function.

File: velox/substrait/tests/VeloxSubstraitRoundTripTest.cpp (hunk 12 above)
Comment (mbasmanova) [reply velox/substrait/tests/VeloxSubstraitRoundTripTest.cpp]: > Added TODO in this PR. More debugging is needed for the conversion from string to expression lists in DuckDB parser.

Sounds good. Would you file an issue to explain what the problem is and maybe provide a small repro?

File: velox/duckdb/conversion/DuckParser.h [hunk 14]
```diff
@@ -30,6 +30,9 @@ struct ParseOptions {
   // Retain legacy behavior by default.
   bool parseDecimalAsDouble = true;
   bool parseIntegerAsBigint = true;
+
+  // The prefix of registered functions.
+  std::string functionPrefix = "";
```
Comment (mbasmanova) [inline velox/duckdb/conversion/DuckParser.h:35]: I'm not sure I understand this change. Would you clarify. Changes to DuckParser need to be submitted in a separate PR with a test.

File: velox/duckdb/conversion/DuckParser.cpp [hunk 15]
```diff
@@ -378,7 +378,8 @@ std::shared_ptr<const core::IExpr> parseOperatorExpr(
     params.emplace_back(parseExpr(*operExpr.children[0], options));
     params.emplace_back(std::make_shared<const core::ConstantExpr>(
         ARRAY(valueType), variant::array(values), std::nullopt));
-    auto inExpr = callExpr("in", std::move(params), getAlias(expr));
+    auto inExpr = callExpr(
+        options.functionPrefix + "in", std::move(params), getAlias(expr));
```
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/duckdb/conversion/DuckParser.cpp:382]: Prefix was also applied to other scalar functions.
https://github.com/facebookincubator/velox/pull/4690/files#diff-9a83c319e87e7171bc95fdd1ccb3f2af4c19e6f3a7048c095f4238a955a74d1dR228

File: velox/duckdb/conversion/DuckParser.h (hunk 14 above)
Comment (rui-mo) [reply velox/duckdb/conversion/DuckParser.h:35]: Previously, DuckParser generates `in` by default which goes into presto sql function, when spark sql functions were registered with prefix `spark`. Here, I added the prefix as an option to let the user can control the generated function.

File: velox/duckdb/conversion/DuckParser.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/duckdb/conversion/DuckParser.cpp:382]: Yes, only `in` was changed to ensure this PR is small. Maybe other functions' change should be put in another PR if needed. How do you think?

File: velox/substrait/VeloxToSubstraitExpr.cpp (hunk 4 above)
Comment (rui-mo) [reply velox/substrait/VeloxToSubstraitExpr.cpp]: I find presto and spark both register it as `in`, seeing [GeneralFunctionsRegistration.cpp](https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/registration/GeneralFunctionsRegistration.cpp#L24) and [In.cpp](https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/In.cpp#L117). How about using certain definition just like aggregate names in [AggregateNames.h](https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/aggregates/AggregateNames.h)?

File: velox/duckdb/conversion/DuckParser.h (hunk 14 above)
Comment (rui-mo) [reply velox/duckdb/conversion/DuckParser.h:35]: Opened https://github.com/facebookincubator/velox/pull/5212 for this change. If the idea is OK, I can apply function prefix on all the needed places in DuckParser.

File: velox/substrait/VeloxToSubstraitPlan.h [hunk 16]
```diff
@@ -36,11 +36,13 @@ class VeloxToSubstraitPlanConvertor {
   /// Convert Velox PlanNode into Substrait Plan.
   /// @param vPlan Velox query plan to convert.
   /// @param arena Arena to use for allocating Substrait plan objects.
+  /// @param prefix Registered function prefix.
```
Comment (mbasmanova) [inline velox/substrait/VeloxToSubstraitPlan.h:39]: Would you clarify what it is used for / applied to?

File: velox/substrait/SubstraitToVeloxPlan.h [hunk 17]
```diff
@@ -80,7 +80,10 @@ class SubstraitVeloxPlanConverter {
   core::PlanNodePtr toVeloxPlan(const ::substrait::SortRel& sortRel);
 
   /// Convert Substrait Plan into Velox PlanNode.
-  core::PlanNodePtr toVeloxPlan(const ::substrait::Plan& substraitPlan);
+  /// Prefix: the prefix of registered functions.
+  core::PlanNodePtr toVeloxPlan(
+      const ::substrait::Plan& substraitPlan,
+      const std::string& prefix = "");
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxPlan.h:86]: Shouldn't the prefix be passed to the constructor?

File: velox/substrait/SubstraitToVeloxExpr.h [hunk 18]
```diff
@@ -63,6 +64,22 @@ class SubstraitVeloxExprConverter {
...
+      const RowTypePtr& inputType);
+
+  /// Create a constant expression from constant expressions by wrapping into
+  /// array.
+  std::shared_ptr<const core::ConstantTypedExpr> toConstantExpr(
+      const std::vector<std::shared_ptr<const core::ConstantTypedExpr>>&
+          constantExprs);
+
+  /// Return the function prefix.
+  std::string getFuncPrefix() {
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxExpr.h:79]: use const& for return value and make this method const

**Review by mbasmanova — COMMENTED** [review]
@rui-mo Looks good to me % a few comments.

File: velox/substrait/SubstraitToVeloxPlan.h (hunk 17 above)
Comment (rui-mo) [reply velox/substrait/SubstraitToVeloxPlan.h]: Moved to the constructor.

File: velox/substrait/SubstraitToVeloxExpr.h (hunk 18 above)
Comment (rui-mo) [reply velox/substrait/SubstraitToVeloxExpr.h]: Fixed.

File: velox/substrait/VeloxToSubstraitPlan.h (hunk 16 above)
Comment (rui-mo) [reply velox/substrait/VeloxToSubstraitPlan.h]: Added more info for it.

**Comment by rui-mo** [comment]
> @rui-mo Looks good to me % a few comments.

@mbasmanova Fixed above comments. Could you help review again? Thank you.

File: velox/substrait/SubstraitParser.h [hunk 19]
```diff
@@ -72,9 +72,11 @@ class SubstraitParser {
 
   /// Find the Velox function name according to the function id
   /// from a pre-constructed function map.
+  /// prefix: the prefix of registered SQL functions.
   std::string findVeloxFunction(
       const std::unordered_map<uint64_t, std::string>& functionMap,
-      uint64_t id) const;
+      uint64_t id,
+      const std::string& prefix = "") const;
```
Comment (mbasmanova) [inline velox/substrait/SubstraitParser.h:79]: Do we need a default value here? If not, let's remove.

File: velox/substrait/SubstraitParser.h [hunk 20]
```diff
@@ -72,9 +72,11 @@ class SubstraitParser {
 
   /// Find the Velox function name according to the function id
   /// from a pre-constructed function map.
+  /// prefix: the prefix of registered SQL functions.
```
Comment (mbasmanova) [inline velox/substrait/SubstraitParser.h:75]: @param prefix ...

**Review by mbasmanova — APPROVED** [review]
@rui-mo Looks good to me.

File: velox/substrait/SubstraitParser.h (hunk 19 above)
Comment (rui-mo) [reply velox/substrait/SubstraitParser.h]: Removed.

File: velox/substrait/SubstraitParser.h (hunk 20 above)
Comment (rui-mo) [reply velox/substrait/SubstraitParser.h]: Fixed.

**Comment by FelixYBW** [comment]
@rui-mo  let's close this one to clear up the pending PRs
